src/python/pants/engine/legacy/structs.py

```diff
@@ -302,6 +302,18 @@ def get_sources(self) -> "GlobsWithConjunction":
     return GlobsWithConjunction.for_literal_files(['register.py'], self.address.spec_path)
 
 
+def parse_glob_includes_and_excludes(globs: Iterable[str]) -> Tuple[List[str], List[str]]:
+  """Split up raw glob patterns into include patterns and exclude patterns."""
+  include_patterns: List[str] = []
+  exclude_patterns: List[str] = []
+  for glob in globs:
+    if glob.startswith("!"):
```

Comment:
Could we possibly centralize where the string processing with `!` happens somewhere (`strutil.py`?) so we don't have the BUILD file and command line syntax accidentally drifting away from each other? That might allow reducing the size of the spec tests you've added too, if we can test that part separately.

Reply:
That's why I put this method as a free function rather than a method on `BaseGlobs`. Both the BUILD file and command line use this function to interpret includes vs. excludes, rather than duplicating the check.

Possibly, this function should move to another file but I kept it here for now because our parsing logic lives here. Maybe it should move to `source/filespec.py` or `source/wrapped_globs.py`? I don't think `strutil.py` is appropriate because this is only relevant to filespecs, not generic strings.

Reply:
> That's why I put this method as a free function rather than a method on BaseGlobs

I had PR tunnel vision! Your approach here sounds great!

Reply:
>  I don't think strutil.py is appropriate because this is only relevant to filespecs, not generic strings.

+1!